editor/src/clj/editor/code/data.clj

```diff
@@ -1513,23 +1515,28 @@
   (let [changes (::changes (meta new-lines))]
     (when (and changes (= (hash new-lines) (:current-version (peek changes))))
       (let [old-version (hash old-lines)
-            ret (into
-                  []
-                  (comp
-                    (drop-while #(not= old-version (:previous-version %)))
-                    (map :ascending-cursor-ranges-and-replacements)
-                    (mapcat
-                      (fn [ascending-cursor-ranges-and-replacements]
-                        (eduction
-                          (map (fn [[cursor-range replacement]]
-                                 [(->CursorRange
-                                    (cursor-range-start cursor-range)
-                                    (cursor-range-end cursor-range))
-                                  replacement]))
-                          (reverse ascending-cursor-ranges-and-replacements)))))
-                  changes)]
-        (when (pos? (count ret))
-          ret)))))
+            last-version-index (loop [i (dec (count changes))]
+                                 (cond
+                                   (neg? i) i
+                                   (= old-version (:previous-version (changes i))) i
+                                   :else (recur (dec i))))]
+        (when-not (neg? last-version-index)
+          (let [ret (into
+                      []
+                      (comp
+                        (map :ascending-cursor-ranges-and-replacements)
+                        (mapcat
+                          (fn [ascending-cursor-ranges-and-replacements]
+                            (eduction
+                              (map (fn [[cursor-range replacement]]
+                                     [(->CursorRange
+                                        (cursor-range-start cursor-range)
+                                        (cursor-range-end cursor-range))
+                                      replacement]))
+                              (reverse ascending-cursor-ranges-and-replacements)))))
+                      (subvec changes last-version-index))]
+            (when (pos? (count ret))
+              ret)))))))
```

Comment:
Minor improvement: we look for incremental diff from the end of the change log now so we don't send unnecessary back-and-forth diffs.